test/libyul/objectCompiler/subobject_data_access.yul

```diff
@@ -0,0 +1,105 @@
+object "A" {
+    code {
+        sstore(0, datasize("A"))
+        sstore(1, datasize("dataA1"))
+        sstore(2, datasize("dataA2"))
+        pop(dataoffset("B"))
+        pop(datasize("B.dataB1"))
+        sstore(3, calldataload(0))
+        pop(dataoffset("B.C"))
+        pop(datasize("B.C.dataC1"))
+        sstore(4, calldataload(0))
+    }
```

Comment:
When it's so irregular and bunched up it's a bit hard to verify the output. The first thing I'd like to see is for example that something like `A.dataA1` gives the same size as `dataA1` used inside `A`. Another is that `B.dataB1` does not have the same size as `B`. I can't see these things here. `calldataload()` also does not seem important. My suggestion would be to split it into 3 simpler tests:

- Qualified vs unqualified data access
    ```yul
    object "A" {
        code {
            sstore(0, datasize("B"))
            sstore(1, datasize("B.data1"))
            sstore(2, datasize("B.data2"))
        }

        object "B" {
            code {
                sstore(0, datasize("B"))
                sstore(1, datasize("data1"))
                sstore(2, datasize("data2"))
            }

            data "data1" "abcdef"
            data "data2" "123"
        }
    }
    ```
- Qualified access inside the same object
    ```yul
    object "A" {
        code {
            sstore(0, datasize("A"))
            sstore(1, datasize("A.data1"))
            sstore(2, datasize("A.data2"))
        }

            data "data1" "abcdef"
            data "data2" "123"
    }
    ```
- Something more nested (like your `A.B.C` example), may include `dataoffset()`, etc.

Reply:
@cameel sorry for the delay on this PR. I have a question regarding the second test you have exampled.
```
object "A" {
    code {
        sstore(0, datasize("A"))
        sstore(1, datasize("A.data1"))
        sstore(2, datasize("A.data2"))
    }

        data "data1" "abcdef"
        data "data2" "123"
}
```

In the current behavior this test wouldn't be compiled because here

https://github.com/ethereum/solidity/blob/591df042115c6df190faa26a1fb87617f7772db3/libyul/ObjectParser.cpp#L215

when data objects are being parsed it considers exact names `data1` `data2` etc.

however, when this data object is being tried to get from parsed object according to the `funCall.arguments` in

https://github.com/ethereum/solidity/blob/591df042115c6df190faa26a1fb87617f7772db3/libyul/AsmAnalysis.cpp#L395

the condition fails because of hash mismatch.

So my question is whether this case should be handled or this test should not work?
If the answer is the first option how about checking if an argument start with `this` object name and ignore this suffix, or when an object is being parsed add refs to data objects for both cases: qualified and unqualified calls. For chunks (nested objects everything works good).

Reply:
Oh, so he fact that accessing the object by its name is just a special case? I did not expect that.

This case is not actually listed in the bug (and it does not trigger an ICE) so we could do it either way. On one hand it seems that this is a pretty natural extension of the syntax and should work. On the other, I now see that this would result in objects having more than one path. For that reason maybe let's not change it here - let's keep it producing an error. But please still add this case as a syntax test.

> If the answer is the first option how about checking if an argument start with `this` object name and ignore this suffix, or when an object is being parsed add refs to data objects for both cases: qualified and unqualified calls.

If we were to implement this, I think the approach with removing the prefix would be better than having duplicated entries.

But as I said, for now it's fine as an error.

Reply:
Test added, #13807 was added not to forget it despite low priority and probably will not be implemented, just for history.